### ISSUE TITLE
Simplify way users have access to grid metrics

### ISSUE THREAD

**Issue by navidcy** [CliMA/Oceananigans.jl#2058]
We should figure out an interface to help users access grid metrics. We shouldn't have to write things like `grid.Δxᵃᵃᶜ` in the examples.

_Originally posted by @glwagner in https://github.com/CliMA/Oceananigans.jl/issues/2050#issuecomment-967266791_

**Comment by navidcy** [comment]
We need a function like `dx = grid_spacing(grid, 1)` or something which returns a number (if constant) or a field (if not).

_Originally posted by @glwagner in https://github.com/CliMA/Oceananigans.jl/issues/2050#issuecomment-967748805_

**Comment by tomchor** [comment]
> We need a function like `dx = grid_spacing(grid, 1)` or something which returns a number (if constant) or a field (if not).
> 
> _Originally posted by @glwagner in [#2050 (comment)](https://github.com/CliMA/Oceananigans.jl/pull/2050#issuecomment-967748805)_

Wouldn't it be better to have something like `Δx = grid_spacing(model.velocities.v, 1)`? Because I'm assuming the `Δx` has to change depending if they want to use it on a quantity that lies on a `Face` or `Center` node in that direction. Maybe just `Δx = grid_spacing(Face, grid, 1)`, or something akin to the already-existing `nodes` function, that returns all 3 at the same time?

**Comment by glwagner** [comment]
That's correct that this function needs to specify location.

**Comment by navidcy** [comment]
also `, 1)` means nothing... better to be `, :x)`a

**Comment by glwagner** [comment]
It's mirroring Base julia, eg

```julia
julia> a = rand(2, 3, 4)
2×3×4 Array{Float64, 3}:
[:, :, 1] =
 0.264469  0.977549  0.937297
 0.491955  0.710279  0.315507

[:, :, 2] =
 0.243796  0.783644  0.820073
 0.682626  0.351825  0.503396

[:, :, 3] =
 0.726912  0.287171  0.671998
 0.234873  0.993365  0.913062

[:, :, 4] =
 0.507489  0.568977  0.327032
 0.59475   0.547478  0.459589

julia> size(a, 3)
4julia> a = rand(2, 3, 4)
2×3×4 Array{Float64, 3}:
[:, :, 1] =
 0.264469  0.977549  0.937297
 0.491955  0.710279  0.315507

[:, :, 2] =
 0.243796  0.783644  0.820073
 0.682626  0.351825  0.503396

[:, :, 3] =
 0.726912  0.287171  0.671998
 0.234873  0.993365  0.913062

[:, :, 4] =
 0.507489  0.568977  0.327032
 0.59475   0.547478  0.459589

julia> size(a, 3)
4
```

Having named dimensions is a cool idea, but I think we should do this properly (eg #1700) rather than with an adhoc implementation.

**Comment by aramirezreyes** [comment]
Just here to say that I had to update some code from using `grid.xC` to using `grid.Δxᵃᵃᶜ` which is much uglier to read and  more difficult to write, so I support a redesign or some more user-friendly way.

**Comment by glwagner** [comment]
I propose

```julia
xspacing(X, Y, Z, grid)
yspacing(X, Y, Z, grid)
zspacing(X, Y, Z, grid)
```

where `X, Y, Z` are all _instantiated_ locations eg

```julia
x = xspacing(Center(), Center(), Center(), grid)
```

I propose we avoid "one location" versions that only work for rectilinear / lat-lon grids.

**Comment by tomchor** [comment]
> I propose
> 
> ```julia
> xspacing(X, Y, Z, grid)
> yspacing(X, Y, Z, grid)
> zspacing(X, Y, Z, grid)
> ```
> 
> where `X, Y, Z` are all _instantiated_ locations eg
> 
> ```julia
> x = xspacing(Center(), Center(), Center(), grid)
> ```
> 
> I propose we avoid "one location" versions that only work for rectilinear / lat-lon grids.

I like this API. Although, since this is a user-facing function, I'd vote for us to have one-location methods defined only for grids that support it. Although I don't feel strongly about it.

**Comment by glwagner** [comment]
I'm wondering if we should abandon the one-location methods. They are convenient, but don't generalize well. They may not even generalize to all immersed boundary methods; for example even with simple `GridFittedBottom` the vertical grid spacing depends on horizontal locations.

My thought is that the overall system is simpler and we avoid "dialects" if we are disciplined and always require all three locations.

**Comment by navidcy** [comment]
+1 on this

**Comment by glwagner** [comment]
Tweak on the above proposal:

Let's build a system of functions:

```julia
xnode
xnodes

xspacing
xspacings

xarea
xareas
```

and similar for `y, z`, where these are called with

```julia
xnode(i, j, k, grid, X, Y, Z)
xnodes(grid, X, Y, Z)

xspacing(i, j, k, grid, X, Y, Z)
xspacings(grid, X ,Y, Z)
```

etc, where for example `X=Y=Z=Center()`. `xnode` and `xspacing` return values at `i, j, k`; while `xnodes` and `xspacings` return `Field`.

This proposes a few changes:

* The "location arguments" are moved to the end of the function signatures. This matters for the "singleton" functions `xnode`, etc, because it means we can compose arguments:

```julia
δxᶠᶜᶜ(i, j, k, grid, xnode, X, Y, Z)
```

Then, following the same pattern for `xnodes` preserves the mneumonic and muscle memory; ie we expect to write `X, Y, Z` after `grid`.

* If we want one-location flavors for the `xnodes`, etc, we can define them

```julia
xnodes(grid, X)
```

For the reasons stated above, I'm worried this does a disservice to the community since user code is less general therefore less generally applicable. But perhaps the gain in convenience is worth it.

* Since the global values are now `Field`, obtaining arrays for plotting requires `interior(xnodes(grid, X, Y, Z))`. I think `Field` is better because then coordinates can take part in abstract operations, and things will likely work better for complicated grids like cubed sphere / MultiRegion

**Comment by navidcy** [comment]
We can add or suggest a plot method overload eg `heatmap(a::Field, b::Field, c::Field, args…; kwargs…) = heatmap(interior(a), interior(b), interior(c), args…; kwargs…)`.

**Comment by tomchor** [comment]
> ```julia
> xnode(i, j, k, grid, X, Y, Z)
> xnodes(grid, X, Y, Z)
> 
> xspacing(i, j, k, grid, X, Y, Z)
> xspacings(grid, X ,Y, Z)
> ```
>   
> etc, where for example X=Y=Z=Center(). xnode and xspacing return values at i, j, k; while xnodes and xspacings return Field.

I really like this suggestion, but wouldn't `xpacings(grid, X, Y, Z)` overlap in functionality with `AbstractOperations.Δx`? Would we want to keep both?

**Comment by glwagner** [comment]
`AbstractOperations.Δx` isn't a Field. So the difference would be that `xspacings` collects / calculates the grid spacings, while `AbstractOperations.Δx` is just a placeholder object meant to be used in operations.

But we could maybe come up with a design that conflates these two ideas.

**Comment by glwagner** [comment]
Closed by #2842
